design/design-doc-resource-claim-definition.md

```diff
@@ -0,0 +1,174 @@
+---
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceBindingDefinition
+metadata:
+  name: mysqlinstance-cloudsqlinstance
+spec:
+  claim:
+    apiVersion: database.crossplane.io/v1alpha1
+    kind: MySQLInstance
+  resources:
+  - resource:
+      apiVersion: database.azure.crossplane.io/v1beta1
+      kind: MySQLServer
+    class:
+      apiVersion: database.azure.crossplane.io/v1beta1
+      kind: SQLServerClass
+      strategies:
+        selection:
+        - match
+        match:
+        - fromClaimField: ".spec.engineVersion"
+          toClassField: ".specTemplate.forProvider.version"
+          transforms:
+            toupper: true
+  - resource:
+      apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRule
+    class:
+      apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRuleClass
+```
+
+`ResourceClaimDefinition` is an abstraction that provides guard rails and
+opinions around the process of authoring a `CustomResourceDefinition` that is
+intended for use as a resource claim. Crossplane resource claims are of many
+different kinds, but all have certain things in common. These include:
+
+* Declaring use of the status subresource.
+* Supporting standard Kubernetes object and type metadata.
+* Supporting a binding phase and conditioned status in their `.status`.
+* Supporting a class selector, class reference, resource reference, and
+  connection secret reference in their `.spec`.
+* Supporting a standard set of kubectl printer columns.
+
+A `ResourceClaimDefinition` allows an infrastructure operator to define a
+resource claim without having to specify all of these things correctly. They can
+instead focus on what they're concerned with - defining the purpose specific
+`.spec` fields of their claim. A `ResourceClaimDefinition` controller will watch
+for `ResourceClaimDefinition` instances, filling in the blanks to create a
+corresponding `CustomResourceDefinition`.
+
+`ResourceBindingDefinition` maps a kind of resource claim to the kinds of
+managed resource (and class) that may satisfy it. In addition to declaring the
+relationship between a claim kind and one or more managed resource kinds, a
+`ResourceBindingDefinition` defines the _strategies_ that should be used when
+dynamically provisioning a set of managed resources to satisfy a claim. There
+are two types of strategy:
+
+* _Selection strategies_ constrain which resource classes can satisfy a
+  resource claim, for example by requiring that a particular claim field match
+  a particular resource class field, perhaps after a transform is applied.
+* _Provisioning strategies_ configure how dynamic provisioning is handled, i.e.
+  how to form a managed resource from a claim and a class. One example might be
+  using the resource class's `.specTemplate` as an overlay on the managed
+  resource's `.spec`.
+
+Random notes:
+
+* It's possible to use `ResourceBindingDefinition` without using
```

Comment:
I'm not sure whether this is a good or bad thing. Ability to disturb what `KubernetesCluster` claim does via another resource seems like an open door for surprises for the users of the claims.

Reply:
I would imagine that would depend on whether the user's mental model was "the behaviour of this claim is defined by its RCD" or "the behaviour of this claim is defined by its RCD and RBDs".